src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala

```diff
@@ -864,125 +887,121 @@ class ILoop(defaultIn: BufferedReader = null, protected val out: JPrintWriter =
         }
       }
     }
-    def startup(): Option[String] = withSuppressedSettings {
-      val batchMode = batchText.nonEmpty
-      if (!batchMode) printWelcome()
 
-      in =
-        if (batchMode) SimpleReader(batchText)
-        else if (defaultIn != null) SimpleReader(defaultIn, out, interactive = true)
-        else if (settings.Xnojline || Properties.isEmacsShell) SimpleReader()
-        else new jline.InteractiveReader()
+    // let them start typing, using the splash reader (which avoids tab completion)
+    val firstLine =
+      SplashLoop.readLine(in, prompt) {
+        val ss = this.settings
+        import ss._
 
-      // let them start typing, using the splash reader (which avoids tab completion)
-      val splash = new SplashLoop(in, prompt)
-      splash.start()
+        // wait until after startup to enable noisy settings
+        val saved = {
+          val noisy = List(Xprint, Ytyperdebug, browse)
+          val noisesome = noisy.exists(!_.isDefault)
 
-      // while we go fire up the REPL
-      try {
-        // don't allow ancient sbt to hijack the reader
-        savingReader {
-          createInterpreter()
-          //for (rs <- runnerSettings if rs.execute.isSetByUser) intp.printResults = false
+          if (!isReplDebug && noisesome) Some(Xprint.value, Ytyperdebug.value, browse.value)
+          else None
         }
-        intp.initializeSynchronous()
-        globalFuture = Future successful true
-        if (intp.reporter.hasErrors) {
-          echo("Interpreter encountered errors during initialization!")
-          null
-        } else {
-          loopPostInit()
-          splash.line // blocks
+
+        if (saved.nonEmpty) {
+          this.settings.Xprint.value = List.empty
+          this.settings.browse.value = List.empty
+          this.settings.Ytyperdebug.value = false
         }
-      } finally {
-        splash.stop()
-        // scala/bug#7418 Now, and only now, can we enable TAB completion.
-        if (!(settings.noCompletion || batchMode))
-          in.initCompletion(new ReplCompletion(intp))
-      }
-    }
-    this.settings = settings
 
+        try {
+          createInterpreter()
 
-    try
-      startup() match {
-        case None =>
-          printShellInterrupt()
-          false
-        case Some(line) =>
-          try loop(line) match {
-            case LineResults.EOF if in.interactive => printShellInterrupt()
+          intp.initializeSynchronous()
+          interpreterInitialized.countDown()
+
+          if (intp.reporter.hasErrors) {
+            echo("Interpreter encountered errors during initialization!")
+            throw new InterruptedException
+          }
+
+          // scala/bug#7418 Now that the interpreter is initialized, and `interpretPreamble` has populated the symbol table,
+          // enable TAB completion (we do this before blocking on input from the splash loop,
+          // so that it can offer tab completion as soon as we're ready).
```

Comment:
I don't understand what exactly changed, in 2.12.2 tab-completion works on the first line when waiting for a bit (until the compiler is initialized).